src/main/java/com/xiaomi/infra/pegasus/rpc/async/ReplicaSession.java

```diff
@@ -366,6 +370,26 @@ public void run() {
         TimeUnit.MILLISECONDS);
   }
 
+  // return value:
+  //   true  - pend succeed
+  //   false - pend failed
+  public boolean tryPendRequest(RequestEntry entry) {
+    // double check. the first one doesn't lock the lock.
+    // Because negotiationSucceed only transfered from false to true.
+    // So if it is true now, it will not change in the later.
+    // But if it is false now, maybe it will change soon. So we should use lock to protect it.
+    if (!this.negotiationSucceed) {
+      synchronized (negotiationPendingSend) {
+        if (!this.negotiationSucceed) {
+          negotiationPendingSend.offer(entry);
```

Comment:
Why not use `pendingSend` but create new pending object?

Reply:
We can't use `pendingSend` here.  Because  if the session is connected and the negotiation is not succeed, the session will send the pending messages in `pendingSend`. So if we put it into `pendingSend`, it will procude error.